src/wh_server_keystore.c

```diff
@@ -0,0 +1,293 @@
+        meta->id = searchId;
+        ret = 0;
+    }
+#endif
+    return ret;
+#endif
+}
+
+int hsmEvictKey(whServerContext* server, uint16_t keyId)
+{
+    int ret = 0;
+    int i;
+    /* find key */
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* mark key as erased */
+        if (server->cache[i].meta->id == keyId) {
+            server->cache[i].meta->id = WOLFHSM_ID_ERASED;
+            break;
+        }
+    }
+    /* if the key wasn't found return an error */
+    if (i >= WOLFHSM_NUM_RAMKEYS)
+        ret = BAD_FUNC_ARG;
+    return ret;
+}
+
+int _wh_Server_HandleKeyRequest(whServerContext* server,
+        uint16_t magic, uint16_t action, uint16_t seq,
+        uint8_t* data, uint16_t* size)
+{
+    int ret = 0;
+    uint8_t* in;
+    uint8_t* out;
+    whPacket* packet = (whPacket*)data;
+    whNvmMetadata meta[1] = {0};
+    switch (action)
+    {
+    case WH_KEY_CACHE:
+        /* in is after fixed size fields */
+        in = (uint8_t*)(&packet->keyCacheReq + 1);
+        /* set the metadata fields */
+        meta->id = packet->keyCacheReq.id;
+        meta->flags = packet->keyCacheReq.flags;
+        meta->len = packet->keyCacheReq.len;
+        XMEMCPY(meta->label, packet->keyCacheReq.label, WOLFHSM_NVM_LABEL_LEN);
+        /* get a new id if one wasn't provided */
+        if (meta->id == WOLFHSM_ID_ERASED) {
+            ret =  hsmGetUniqueId(server);
+            if (ret > 0) {
+                meta->id = ret;
+                ret = 0;
+            }
+        }
+        /* write the key */
+        if (ret == 0)
+            ret = hsmCacheKey(server, meta, in);
+        if (ret == 0) {
+            packet->keyCacheRes.id = meta->id;
+            *size = WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyCacheRes);
+        }
+        break;
+    case WH_KEY_EVICT:
+        ret = hsmEvictKey(server, packet->keyEvictReq.id);
+        if (ret == 0) {
+            packet->keyEvictRes.ok = 0;
+            *size = WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyEvictRes);
+        }
+        break;
+#if 0
+    case WOLFHSM_KEY_COMMIT:
+        /* commit the cached key */
+        ret = hsmCommitKey(ctx, packet->keyCommitReq.id);
+        if (ret > 0) {
+            packet->len = sizeof(packet->keyCommitRes);
+            packet->keyCommitRes.ok = 0;
+            ret = 0;
+        }
+        break;
+#endif
+    case WH_KEY_EXPORT:
+        /* out is after fixed size fields */
+        out = (uint8_t*)(&packet->keyExportRes + 1);
+        /* set the metadata fields */
+        meta->id = packet->keyExportReq.id;
+        meta->len = WH_COMM_MTU -
+            (WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->keyExportRes));
+        /* read the key */
+        ret = hsmReadKey(server, meta, out);
+        if (ret == 0) {
+            /* set key len */
+            packet->keyExportRes.len = meta->len;
+            /* set label */
+            XMEMCPY(packet->keyExportRes.label, meta->label,
+                sizeof(meta->label));
```

Comment:
should support variable sized labels

Reply:
Specifically, the caller may not have a full-sized label buffer so we should only copy what the caller can supply and zero-fill the rest.

Reply:
I added the size parameter to `wh_Client_KeyExportResponse` so it will only read out of the packet the length specified